compiler/rustc_data_structures/src/sync.rs

```diff
@@ -303,48 +350,169 @@ cfg_if! {
         use parking_lot::RwLock as InnerRwLock;
 
         use std::thread;
-        pub use rayon::{join, scope};
+
+        #[inline]
+        pub fn join<A, B, RA: DynSend, RB: DynSend>(oper_a: A, oper_b: B) -> (RA, RB)
+        where
+            A: FnOnce() -> RA + DynSend,
+            B: FnOnce() -> RB + DynSend,
+        {
+            if mode::active() {
+                let oper_a = FromDyn::from(oper_a);
+                let oper_b = FromDyn::from(oper_b);
+                let (a, b) = rayon::join(move || FromDyn::from(oper_a.into_inner()()), move || FromDyn::from(oper_b.into_inner()()));
+                (a.into_inner(), b.into_inner())
+            } else {
+                (oper_a(), oper_b())
+            }
+        }
+
+        pub use rayon::scope;
 
         /// Runs a list of blocks in parallel. The first block is executed immediately on
         /// the current thread. Use that for the longest running block.
         #[macro_export]
         macro_rules! parallel {
-            (impl $fblock:tt [$($c:tt,)*] [$block:tt $(, $rest:tt)*]) => {
+            (impl $fblock:block [$($c:expr,)*] [$block:expr $(, $rest:expr)*]) => {
                 parallel!(impl $fblock [$block, $($c,)*] [$($rest),*])
             };
-            (impl $fblock:tt [$($blocks:tt,)*] []) => {
-                ::rustc_data_structures::sync::scope(|s| {
```

Comment:
Can you get rid of the tupling by also wrapping this closure in `FromDyn`?

Reply:
In this way we have to write the macro like this:
```
let spawn_block = rustc_data_structures::sync::FromDyn::from(move |s: &rayon::Scope<$scope> | {
    // ...
});
::rustc_data_structures::sync::scope(|s| spawn_block.into_inner()(s));
```
and pass `'tcx` to macro as the $scope.

I didn't think of a way to avoid it

Reply:
Trying what you methioned that recreating a `scope` function, hope this works